egui/src/widgets/text_edit.rs

```diff
@@ -656,7 +665,18 @@ impl<'t> TextEdit<'t> {
 
         ui.memory().id_data.insert(id, state);
 
-        response.widget_info(|| WidgetInfo::text_edit(&*text));
+        if response.changed {
+            response.widget_info(|| WidgetInfo::text_edit(&*text, &*prev_text));
+        } else if let Some(text_cursor) = text_cursor {
+            response.has_widget_info = true;
+            response.widget_info(|| {
+                WidgetInfo::text_selection_changed(
+                    text_cursor.primary.ccursor.index,
+                    text_cursor.secondary.ccursor.index,
+                    &*text,
+                )
+            });
+        }
```

Comment:
this can be simplified to something like this:

``` rust
if let Some(prev_text) = prev_text {
    let info = WidgetInfo::text_changed(prev_text, text);
    self.ctx.output().events.push(OutputEvent::TextChanged(info));
} else if text_cursor_changed {
    let char_range = text_cursor.primary.ccursor.index..text_cursor.secondary.ccursor.index;
    let info = WidgetInfo::text_cursor_changed(char_range, text);
    self.ctx.output().events.push(OutputEvent::TextCursorChanged(info));
}

// Response::widget_info will emit events if we are clicked, doublt-clicked etc:
response.widget_info(|| WidgetInfo::text_edit(&*text));
```

Reply:
I feel like we should consolidate accessibility event dispatch into as few places as possible, and we should locate the dispatches in the places that make the most sense.

E.g. `Response::widget_info` dispatches a click event when `self.clicked()` is true. That makes a lot of sense and seems maintainable. It currently dispatches `ValueChanged` when `self.changed` is true. Also makes a lot of sense to me. I'm hesitant to have a bunch of different places pushing change events, different places pushing click events, etc. if we can at all avoid it. The fewer choke points for accessibility event dispatch, the better IMO.

That said, something like tracking widget-internal state like the cursor position can absolutely push events directly, so I'll nix `has_widget_info`.

Thoughts?

Reply:
I agree that dispatching all events from one function has a charm to it, and that was my original intent. What break the charm is that we are constructing different content for the `WidgetInfo`:s for different events.

We could have `TextEdit` call `widget_info` just as on `main` but with a `WidgetInfo` that now contains the latest cursor and the before/after text pair (every frame, even if nothing changed). `Response` can then be augmented with the boolean `text_cursor_changed` which `widget_info` would check and dispatch that special event. `OutputEvent::TextChanged` can be removed (`OutputEvent::Changed` would be sent when text changes, and the receiver can figure out that it was text that changed). I think this is the cleanest design.